File: backend/transcend/src/test/java/code/users/logic/ManageFriendsTest.java
package code.users.logic;

import static code.users.domain.model.UserFixtures.AVATAR_NAME_FIXTURE;
import static code.users.domain.model.UserFixtures.AVATAR_URL_FIXTURE;
import static code.users.domain.model.UserFixtures.USER_ID_FIXTURE;
import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import code.users.domain.exceptions.UserNotFoundException;
import code.users.domain.model.FriendId;
import code.users.domain.model.UserDetails;
import code.users.ports.in.ManageFriendsUseCase;
import code.users.ports.out.UserDao;
import java.util.Map;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;

@SpringJUnitConfig(ManageFriendsTest.ManageFriendsTestConfig.class)
@RequiredArgsConstructor(onConstructor_ = {@Autowired})
class ManageFriendsTest {

  @Configuration
  @Import(ManageFriends.class)
  static class ManageFriendsTestConfig {}

  private final ManageFriendsUseCase service;

  @MockitoBean private UserDao userDao;

  @Test
  void addFriendSuccessfully() {
    // given
    var friendId = FriendId.of(UUID.randomUUID());
    when(userDao.exists(friendId)).thenReturn(true);

    // when
    service.addFriend(USER_ID_FIXTURE, friendId);

    // then
    verify(userDao).addFriend(USER_ID_FIXTURE, friendId);
  }

  @Test
  void removeFriendSuccessfully() {
    // given
    var friendId = FriendId.of(UUID.randomUUID());
    when(userDao.exists(friendId)).thenReturn(true);

    // when
    service.removeFriend(USER_ID_FIXTURE, friendId);

    // then
    verify(userDao).removeFriend(USER_ID_FIXTURE, friendId);
  }

  // --- New tests for the updated logic ---

  @Test
  void addFriendThrowsExceptionWhenFriendDoesNotExist() {
    // given
    var friendId = FriendId.of(UUID.randomUUID());
    when(userDao.exists(friendId)).thenReturn(false);

    // when & then
    assertThrows(UserNotFoundException.class, () -> service.addFriend(USER_ID_FIXTURE, friendId));
    verify(userDao, never()).addFriend(any(), any());
  }

  @Test
  void removeFriendThrowsExceptionWhenFriendDoesNotExist() {
    // given
    var friendId = FriendId.of(UUID.randomUUID());
    when(userDao.exists(friendId)).thenReturn(false);

    // when & then
    assertThrows(
        UserNotFoundException.class, () -> service.removeFriend(USER_ID_FIXTURE, friendId));
    verify(userDao, never()).removeFriend(any(), any());
  }

  @Test
  void getFriendListSuccessfully() {
    // given
    var expectedFriends =
        Map.of(
            FriendId.of(UUID.randomUUID()),
            UserDetails.builder()
                .displayName(AVATAR_NAME_FIXTURE)
                .avatarUrl(AVATAR_URL_FIXTURE)
                .build());
    when(userDao.getFriendList(USER_ID_FIXTURE, 0, 10)).thenReturn(expectedFriends);

    // when
    var result = service.getFriendList(USER_ID_FIXTURE, 0, 10);

    // then
    verify(userDao).getFriendList(USER_ID_FIXTURE, 0, 10);
    assertThat(result).hasSize(1);
    assertThat(result.values().iterator().next().getDisplayName()).isEqualTo(AVATAR_NAME_FIXTURE);
  }
}

File: backend/transcend/src/test/kotlin/code/users/logic/ManageFriendsTest.kt
package code.users.logic

import code.users.bootstrap.FriendDaoTestSupport
import code.users.domain.exceptions.UserNotFoundException
import code.users.domain.model.FriendFixtures.*
import code.users.domain.model.UserFixtures.USER_ID_FIXTURE
import code.users.ports.`in`.ManageFriendsUseCase
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.matchers.maps.shouldHaveSize
import io.kotest.matchers.shouldBe
import org.springframework.context.annotation.Import

@Import(
  ManageFriends::class
)
class ManageFriendsTest(
  private val service: ManageFriendsUseCase
) : FriendDaoTestSupport() {

  init {
    Given("a user and existing potential friends") {

      When("the add friend service is executed for friend 1") {
        service.addFriend(USER_ID_FIXTURE, FRIEND1_CLASS_ID_FIXTURE)

        Then("friend 1 should be in the user's friend list") {
          val friends = userDao.getFriendList(USER_ID_FIXTURE, 0, 10)
          friends.containsKey(FRIEND1_CLASS_ID_FIXTURE) shouldBe true
        }
      }
    }

    Given("the user is already friends with friend 2") {
      userDao.addFriend(USER_ID_FIXTURE, FRIEND2_CLASS_ID_FIXTURE)

      When("the remove friend service is executed for friend 2") {
        service.removeFriend(USER_ID_FIXTURE, FRIEND2_CLASS_ID_FIXTURE)

        Then("friend 2 should be removed from the database") {
          val friends = userDao.getFriendList(USER_ID_FIXTURE, 0, 10)
          friends.containsKey(FRIEND2_CLASS_ID_FIXTURE) shouldBe false
        }
      }
    }

    Given("the system is queried for a friend list") {
      userDao.addFriend(USER_ID_FIXTURE, FRIEND1_CLASS_ID_FIXTURE)
      userDao.addFriend(USER_ID_FIXTURE, FRIEND2_CLASS_ID_FIXTURE)

      When("requesting page 0") {
        val result = service.getFriendList(USER_ID_FIXTURE, 0, 10)

        Then("it should return both friends") {
          result shouldHaveSize 2
          result[FRIEND1_CLASS_ID_FIXTURE]?.displayName shouldBe FRIEND1_NAME_FIXTURE
          result[FRIEND2_CLASS_ID_FIXTURE]?.displayName shouldBe FRIEND2_NAME_FIXTURE
        }
      }
    }

    Given("a completely random non-existent ID") {
      When("attempting to add the friend") {
        Then("it should throw UserNotFoundException") {
          shouldThrow<UserNotFoundException> {
            service.addFriend(USER_ID_FIXTURE, NON_EXISTENT_FRIEND)
          }
        }
      }
    }
  }
}